agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/exporter/utils/SanitizationHelper.java

```diff
@@ -0,0 +1,143 @@
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+package com.microsoft.applicationinsights.agent.internal.exporter.utils;
+
+import com.microsoft.applicationinsights.agent.internal.common.Strings;
+import java.util.HashMap;
+import java.util.Map;
+
+public class SanitizationHelper {
+  public static final int MAX_KEY_LENGTH = 150;
+  public static final int MAX_VALUE_LENGTH = 8192;
+  public static final int MAX_NAME_LENGTH = 1024;
+  public static final int MAX_ID_LENGTH = 512;
+  public static final int MAX_MESSAGE_LENGTH = 32768;
+  public static final int MAX_URL_LENGTH = 2048;
+
+  /** Function to sanitize both key and value in properties. */
+  @SuppressWarnings("ReturnsNullCollection")
+  public static Map<String, String> sanitizeProperties(Map<String, String> properties) {
+    if (properties == null) {
+      return null;
+    }
+    if (!needsSanitizing(properties)) {
+      // this is an optimization to avoid any memory allocation in the normal case
+      return properties;
+    }
+    Map<String, String> sanitized = new HashMap<>();
+    for (Map.Entry<String, String> entry : properties.entrySet()) {
+      String sanitizedKey = sanitizeKey(entry.getKey());
+      String sanitizedValue = sanitizeValue(entry.getValue());
+      if (!Strings.isNullOrEmpty(sanitizedValue)) {
+        String uniqueKey = makeKeyUnique(sanitizedKey, sanitized);
+        sanitized.put(uniqueKey, sanitizedValue);
+      }
+    }
+    return sanitized;
+  }
+
+  private static boolean needsSanitizing(Map<String, String> properties) {
+    for (Map.Entry<String, String> entry : properties.entrySet()) {
+      String key = entry.getKey();
+      String value = entry.getValue();
+      if (Strings.isNullOrEmpty(key)
+          || Strings.isNullOrEmpty(value)
+          || key.length() > MAX_KEY_LENGTH
+          || value.length() > MAX_VALUE_LENGTH) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /** Function to create unique key. */
+  private static String makeKeyUnique(String key, Map<String, ?> map) {
+    if (!map.containsKey(key)) {
+      return key;
+    }
+    String truncatedKey = Strings.truncate(key, MAX_KEY_LENGTH - 3);
+    int candidate = 1;
+    do {
+      key = truncatedKey + candidate;
+      ++candidate;
+    } while (map.containsKey(key));
+    return key;
+  }
+
+  /** Function to sanitize value. */
+  private static String sanitizeValue(String value) {
+    return Strings.trimAndTruncate(value, MAX_VALUE_LENGTH);
+  }
+
+  /** Function to sanitize key. */
+  private static String sanitizeKey(String key) {
+    String sanitizedKey = Strings.trimAndTruncate(key, MAX_KEY_LENGTH);
+    return makeKeyNonEmpty(sanitizedKey);
+  }
+
+  /** Function to return non empty key. */
+  private static String makeKeyNonEmpty(String sanitizedKey) {
+    // TODO check if this is valid. Below code is based on .Net implementation of the same
+    return Strings.isNullOrEmpty(sanitizedKey) ? "empty" : sanitizedKey;
```

Comment:
if this happens many times, empty1, empty2, empty3.... empty34 could end up in the payload.. if the key is empty, it should not get added.. sanitizedKey should not get called in that case.  I wouldn't simply copy logic from .net sdk.

Reply:
there can be only one empty key in the properties hashMap

Reply:
know hashMap won't allow duplicate keys.. my suggestion is not to add any key-value pair when the key is empty rather than use "empty" as the key in the hashmap.

Reply:
@kryalama is this behavior copied from .NET SDK?

Reply:
yes this is copied from .Net impelementation

Reply:
ok, let's keep it as is then